Toktok migration

Diadlo opened the pull request:
Migration qtox on [toktok/c-toxcore](https://github.com/TokTok/c-toxcore)
- [x] Group API migration
- [x] Error handling

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/3736)

<!-- Reviewable:end -->

Diadlo committed:
refactor(core, widget): Update toxcore group API (new: conference)

Diadlo committed:
refactor(core): Update toxcore callbacks

Diadlo committed:
refactor(core): Add Core* as userdata to tox_iterate

zetok commented:
needed:
- [x] update travis
- [x] update INSTALL.md – version `=0.1`
- [x] mark as a breaking change
- [x] versioned TokTok toxcore release with ~stable API

Edit: also
- [x] update bootstrap scripts – once toxcore gets stable `v0.1.0` release, `git checkout` it (also in docs)

Diadlo committed:
refactor(core): Added error handling

Diadlo committed:
refactor(core): Remove code duplication for different types of messages

Diadlo committed:
chore(INSTALL.md): Updated links on toxcore repository

BREAKING CHANGE: Changed toxcore public API version to 0.1.0

Diadlo committed:
chore: Updated build scrips

sudden6 commented:
repository in scripts needs to be updated to `toktok/c-toxcore`

initramfs commented:
Reviewed 4 of 9 files at r2.
Review status: 4 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

_[src/core/core.cpp, line 236 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv7i9tmy5rlZ_SFDR_:-KUv7i9tmy5rlZ_SFDRa:b-yeakuv) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L236)):_

> ``` cpp
>  */
> void Core::start()
> {
> ```

Add a print statement to show the version of toxcore in use. This will help debug issues with runtime loading of the toxcore (and to verify that the correct toxcore version is being used).

Something like:

``` c++
qDebug().nospace() << "Using toxcore version " << tox_version_major() << '.' << tox_version_minor() << '.' << tox_version_patch();
```

---

_[src/core/core.cpp, line 529 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv8pGkm1yQdR_4WhNA:-KUv8pGlJBCiU95Xhw4a:bdjh2de) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L529)):_

> ``` cpp
> }
> 
> #if TOX_VERSION_IS_API_COMPATIBLE(0, 0, 1)
> ```

Why not use `TOX_VERSION_IS_API_COMPATIBLE(0, 0, 0)` or simply invert the #if?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)_

<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 4 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

_[src/core/core.cpp, line 236 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv7i9tmy5rlZ_SFDR_:-KUxoPPTtb9D0IhROhSz:blxlibg) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L236)):_
<details><summary>

<i>Previously, initramfs wrote…</i></summary>



> Add a print statement to show the version of toxcore in use. This will help debug issues with runtime loading of the toxcore (and to verify that the correct toxcore version is being used).
> 
> Something like:
> 
> ``` c++
> qDebug().nospace() << "Using toxcore version " << tox_version_major() << '.' << tox_version_minor() << '.' << tox_version_patch();
> ```
> 
> </details>
> We have toxcore version on `About` tab in settings

---

_[src/core/core.cpp, line 529 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv8pGkm1yQdR_4WhNA:-KUxq3bnLXH6_xBWQdkb:b-8wl5mo) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L529)):_
<details><summary>

<i>Previously, initramfs wrote…</i></summary>



> Why not use `TOX_VERSION_IS_API_COMPATIBLE(0, 0, 0)` or simply invert the #if?
> </details>
> Because
> 
> ```
>  #define TOX_VERSION_IS_API_COMPATIBLE(MAJOR, MINOR, PATCH)      \                
>      (TOX_VERSION_MAJOR == MAJOR &&                                \                
>      (TOX_VERSION_MINOR > MINOR ||                                \                
>      (TOX_VERSION_MINOR == MINOR &&                              \                
>      TOX_VERSION_PATCH >= PATCH)))                         
> ```

How inver `#if`?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)_

<!-- Sent from Reviewable.io -->

initramfs commented:
I feel like PR makes is such that once qTox has been linked at compile time to a newer version of toxcore, it won't be able to load older versions of toxcore at runtime (since you're actively avoiding compiling stuff for older toxcore instances). The similar problem occurs the other way around. Whilst this is perfectly fine (and optimal for distributions that compile on the target machine), it may pose an issue where pre-packaged versions of qTox are stored in repositories against a toxcore revision that's different to the one it's been compiled for.

At least in the case of transitioning from irungetoo/toxcore to TokTok/c-toxcore, would it be possible to make a runtime decision between which revision to load for?

---

Review status: 4 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

_[src/core/core.cpp, line 236 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv7i9tmy5rlZ_SFDR_:-KV-Kxl1orAtqIBHHivr:b-on5ytp) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L236)):_
<details><summary>

<i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary>



> We have toxcore version on `About` tab in settings
> </details>
> Fair enough, though I still argue it can be useful in future debugging should bug reports neglect to mention the version of toxcore but this is optional.

---

_[src/core/core.cpp, line 529 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv8pGkm1yQdR_4WhNA:-KV-L9KRXeEz3g7xWV5l:b-pkjrhb) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L529)):_
<details><summary>

<i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary>



> Because
> 
> ```
>  #define TOX_VERSION_IS_API_COMPATIBLE(MAJOR, MINOR, PATCH)      \                
>      (TOX_VERSION_MAJOR == MAJOR &&                                \                
>      (TOX_VERSION_MINOR > MINOR ||                                \                
>      (TOX_VERSION_MINOR == MINOR &&                              \                
>      TOX_VERSION_PATCH >= PATCH)))                         
> ```
> 
> How inver `#if`?
> </details>
> My bad, I read the code wrong. If I read it right this time, you want code to only compile if it's running on toxcore 0.0.0. Well, in that case, you want to check incompatibility with 0.0.1+. As such, invert the if with:

``` c++
#if !TOX_VERSION_IS_API_COMPATIBLE(0, 0, 1)
```

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736#-:-KV-NuLdcGid5XZbHqpA:b-11jbs9)_

<!-- Sent from Reviewable.io -->

zetok commented:
> (…) it may pose an issue where pre-packaged versions of qTox are stored in repositories against a toxcore revision that's different to the one it's been compiled for.
> 
> At least in the case of transitioning from irungetoo/toxcore to TokTok/c-toxcore, would it be possible to make a runtime decision between which revision to load for?

There's no need to do that. As this point, development moved to c-toxcore, and that's what should be packaged.

zetok commented:
There are conflicts, rebase is needed.

iphydf committed:
chore: Update qTox to the latest toxencryptsave API.

Since ownership is somewhat unclear, we now use shared_ptr to pass
these around instead of unique_ptr.

Diadlo committed:
chore(toxcore): Added checkout on v0.1.0 toxcore

Diadlo commented:
Thanks @sudden6 for rebasing

zetok commented:
Reviewed 2 of 4 files at r5.
Review status: 2 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[INSTALL.md, line 44 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9VmjUUPWcxkBPwXq:-KYy9VmjUUPWcxkBPwXr:bgp7i4a) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L44)):*
> ```Markdown
> | [Qt]          | >= 5.3.0    | core, gui, network, opengl, sql, svg, widget, xml |
> | [GCC]/[MinGW] | >= 4.8      | C++11 enabled                                     |
> | [toxcore]     | >= 0.1.0    | core, av                                          |
> ```

Wouldn't it be better to specify `=0.1` ?

---

*[INSTALL.md, line 122 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9SZbK0HuK9brp1M5:-KYy9SZbK0HuK9brp1M6:b-sl5zdk) ([raw file](https://github.com/qtox/qtox/blob/e9778e81a238d2688ba5106b602d08208a9311a9/INSTALL.md#L122)):*
> ```Markdown
> `qtox-alpha`.
> 
> ====
> ```

Why is this removed?

---

*[INSTALL.md, line 345 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9ihYUexp2aQy_hHW:-KYy9ihZC8WggF8HKLyh:bb7phvt) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L345)):*
> ```Markdown
> git clone https://github.com/toktok/c-toxcore.git toxcore
> cd toxcore
> git checkout tags/v0.1.10
> ```

typo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 2 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[INSTALL.md, line 44 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9VmjUUPWcxkBPwXq:-KYyajZFRqc17AZnAKwm:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L44)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

Wouldn't it be better to specify `=0.1` ?
</blockquote></details>

Done.

---

*[INSTALL.md, line 122 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9SZbK0HuK9brp1M5:-KYyaWyghBIhYSl_YEho:b-ijkql9) ([raw file](https://github.com/qtox/qtox/blob/e9778e81a238d2688ba5106b602d08208a9311a9/INSTALL.md#L122)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

Why is this removed?
</blockquote></details>

Missclick in vim :) Done.

---

*[INSTALL.md, line 345 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9ihYUexp2aQy_hHW:-KYyb-cbl5SoLlRk-Ak6:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L345)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 529 at r2](https://reviewable.io:443/reviews/qtox/qtox/3736#-KUv8pGkm1yQdR_4WhNA:-KYybVx6hgC0qprm7N_n:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/dade70a24394804444d29fa0a3f0d70031b47f2a/src/core/core.cpp#L529)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

My bad, I read the code wrong. If I read it right this time, you want code to only compile if it's running on toxcore 0.0.0. Well, in that case, you want to check incompatibility with 0.0.1+. As such, invert the if with:

```c++
#if !TOX_VERSION_IS_API_COMPATIBLE(0, 0, 1)
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

zetok commented:
Review status: 2 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[INSTALL.md, line 44 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9VmjUUPWcxkBPwXq:-KYyflISh7LYK4GxCuvV:b-xsrrms) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L44)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Um, that's `0.1.0`. What I meant is not specifying the patch version.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 2 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[INSTALL.md, line 44 at r6](https://reviewable.io:443/reviews/qtox/qtox/3736#-KYy9VmjUUPWcxkBPwXq:-KYywQiRIsPZnxP6zVAI:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/56c3af0ffc55fbef6aacfa2cad9144df553bc4e8/INSTALL.md#L44)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

Um, that's `0.1.0`. What I meant is not specifying the patch version.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
@zetok found an [issue](https://github.com/qTox/qtox-irc-logs/blob/master/2016/12/%23qtox_20161215.log.txt#L127:L131)

Can't reproduce

zetok commented:
@Diadlo uh, sorry, about that, PEBKAC. I forgot that I've set to pick up calls automatically one one of testing profiles. So there's no issue.

Diadlo commented:
@zetok Lol. Thanks :)

zetok commented:
Reviewed 2 of 8 files at r7.
Review status: 3 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[INSTALL.md, line 632 at r7](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZDPopCl_SHLyUAaL5M:-KZDPopDlQ8muG3HGuQ7:b6pm8xj) ([raw file](https://github.com/qtox/qtox/blob/c1e50f43e26dc30610c4cb715dcc6f4f03773442/INSTALL.md#L632)):*
> ```Markdown
> [Qt]: https://www.qt.io/
> [sqlcipher]: https://www.zetetic.net/sqlcipher/
> [toxcore]: https://github.com/irungentoo/toxcore/
> ```

This too needs to be updated

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 3 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[INSTALL.md, line 632 at r7](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZDPopCl_SHLyUAaL5M:-KZG1m6lw4TOu-nm1jFC:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c1e50f43e26dc30610c4cb715dcc6f4f03773442/INSTALL.md#L632)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

This too needs to be updated
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
Reviewed 1 of 10 files at r3, 7 of 10 files at r4, 6 of 8 files at r7, 1 of 1 files at r8.
Review status: 16 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/persistence/db/rawdatabase.cpp, line 484 at r8](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZGOkkRljOhFek4URz-:-KZGOkkS9e3u7XLLN6I5:bfxox6l) ([raw file](https://github.com/qtox/qtox/blob/792ed34bb5f13f4c26f12aa6c9e5273d21116a35/src/persistence/db/rawdatabase.cpp#L484)):*
> ```cpp
> }
> 
> struct PassKeyDeleter
> ```

docs for this would be nice. Also does `tox_pass_key_free()` take care of overwriting the key?

---

*[src/video/groupnetcamview.cpp, line 263 at r8](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZGPKVsAc6NKeYo8AU2:-KZGPKVsAc6NKeYo8AU3:b-c8nddq) ([raw file](https://github.com/qtox/qtox/blob/792ed34bb5f13f4c26f12aa6c9e5273d21116a35/src/video/groupnetcamview.cpp#L263)):*
> ```cpp
>     Friend* f = FriendList::findFriend(FriendId);
> 
>     for (uint32_t i = 0; i < Core::getInstance()->getGroupNumberPeers(group); ++i)
> ```

this change seems pretty unrelated

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo committed:
docs: Added docs to PassKeyDeleter and updated Core methods

Diadlo commented:
Review status: 16 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/persistence/db/rawdatabase.cpp, line 484 at r8](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZGOkkRljOhFek4URz-:-KZGcPXMKDyEaefEGehl:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/792ed34bb5f13f4c26f12aa6c9e5273d21116a35/src/persistence/db/rawdatabase.cpp#L484)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

docs for this would be nice. Also does `tox_pass_key_free()` take care of overwriting the key?
</blockquote></details>

Done.

---

*[src/video/groupnetcamview.cpp, line 263 at r8](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZGPKVsAc6NKeYo8AU2:-KZG_hKmM9I73id97_VO:bizxonc) ([raw file](https://github.com/qtox/qtox/blob/792ed34bb5f13f4c26f12aa6c9e5273d21116a35/src/video/groupnetcamview.cpp#L263)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

this change seems pretty unrelated
</blockquote></details>

Have no idea, how it got here. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 16 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/video/groupnetcamview.cpp, line 263 at r8](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZGPKVsAc6NKeYo8AU2:-KZHUXxKaPfTUFfJlUIo:b-kggk6r) ([raw file](https://github.com/qtox/qtox/blob/792ed34bb5f13f4c26f12aa6c9e5273d21116a35/src/video/groupnetcamview.cpp#L263)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Have no idea, how it got here. Done.
</blockquote></details>

Oh, not. I canged `getGroupNumberPeers` method. Now it returns `uint32_t`. So change was correct

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

zetok commented:
As far as my testing goes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r8.
Review status: 16 of 20 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736#-:-KZHf4h6USJ6xWKXQbIF:bwtip2m)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
Reviewed 1 of 10 files at r3, 1 of 1 files at r9.
Review status: 18 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/core/core.cpp, line 1063 at r9](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZHoXILwKfqZQW1iHVm:-KZHoXILwKfqZQW1iHVn:b-ip3wn2) ([raw file](https://github.com/qtox/qtox/blob/c69e1a987521adaee14845d9fafa123b35bebd5a/src/core/core.cpp#L1063)):*
> ```cpp
> 
> /**
>  * @brief Return the number of peers in the group chat on success, or -1 on failure
> ```

The comment is now wrong

---

*[src/core/core.cpp, line 1125 at r9](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZHoEB-LP-38l4xYL_O:-KZHoEB0DGwzQgHNthJo:b-h57bgm) ([raw file](https://github.com/qtox/qtox/blob/c69e1a987521adaee14845d9fafa123b35bebd5a/src/core/core.cpp#L1125)):*
> ```cpp
>     }
> 
>     int result = getGroupNumberPeers(groupId);
> ```

`getGroupNumberPeers` no longer returns `int` please update this accordingly, there are probably other occurences that need updates

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 18 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/core/core.cpp, line 1063 at r9](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZHoXILwKfqZQW1iHVm:-KZHtfJ-LwQlTIjbxEI5:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c69e1a987521adaee14845d9fafa123b35bebd5a/src/core/core.cpp#L1063)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

The comment is now wrong
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1125 at r9](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZHoEB-LP-38l4xYL_O:-KZHvezora6ciGNUk2rA:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c69e1a987521adaee14845d9fafa123b35bebd5a/src/core/core.cpp#L1125)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

`getGroupNumberPeers` no longer returns `int` please update this accordingly, there are probably other occurences that need updates
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

zetok commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 18 of 20 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736#-:-KZHwtl256NCtBpFQ_VK:b-uwkowd)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 10 files at r4, 1 of 4 files at r5, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736#-:-KZI-vPK5cVgcoc6E3Q9:b-uwkowd)*
<!-- Sent from Reviewable.io -->

initramfs commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[src/core/core.cpp, line 507 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZISrAux5hg9wq6W1I4:-KZISrAux5hg9wq6W1I5:b-1ed8yr) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L507)):*
> ```cpp
> 
> void Core::onGroupInvite(Tox*, uint32_t friendId, TOX_CONFERENCE_TYPE type,
>                          const uint8_t *data, size_t length, void* _core)
> ```

Pointer aligns to the left.

---

*[src/core/core.cpp, line 507 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIVwtEC0_FgBlyIVjM:-KZIVwtEC0_FgBlyIVjN:b-h7viqc) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L507)):*
> ```cpp
> 
> void Core::onGroupInvite(Tox*, uint32_t friendId, TOX_CONFERENCE_TYPE type,
>                          const uint8_t *data, size_t length, void* _core)
> ```

See point below on reserved identifiers.

---

*[src/core/core.cpp, line 528 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIUULLG1X9vn90Douw:-KZIUULMFnKz75cMWDsX:b5tbwy4) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L528)):*
The use of an identifier who's name starts with an underscore is prohibited in C++:

As per § 17.6.4.3.2:

> Each name that begins with an underscore is reserved to the implementation for use as a name in the
global namespace.

---

*[src/core/core.cpp, line 533 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIY_InHbw9GoJjB2Ww:-KZIY_IodIrRLl0oWiT7:bp2o9ab) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L533)):*
> ```cpp
>     QString message = CString::toString(_message, length);
>     bool isAction;
>     isAction = type == TOX_MESSAGE_TYPE_ACTION;
> ```

```c++
bool isAction;
isAction = type == TOX_MESSAGE_TYPE_ACTION;
```

can be combined into 1 line.

---

*[src/core/core.cpp, line 545 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIVSpn-sarMJQ5fANx:-KZIVSpoS5F_hmUJVXDB:bebjb83) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L545)):*
> ```cpp
> 
> void Core::onGroupTitleChange(Tox*, uint32_t groupId, uint32_t peerId,
>                               const uint8_t* _title, size_t length, void* _core)
> ```

See point above.

---

*[src/core/core.cpp, line 553 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIWAKM9_6hLJrU9sPq:-KZIWAKNKe-MtggSNw5m:b-p63dur) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L553)):*
> ```cpp
> }
> 
> void Core::onReadReceiptCallback(Tox*, uint32_t friendId, uint32_t receipt, void *core)
> ```

You fixed pointer alignment above, but not here. Consider fixing this one too.

---

*[src/core/core.cpp, line 646 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIWF5XxitLTTlNEvBj:-KZIWF5XxitLTTlNEvBk:b-94vet7) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L646)):*
> ```cpp
> }
> 
> void Core::sendGroupMessageWithType(int groupId, const QString &message, TOX_MESSAGE_TYPE type)
> ```

Reference alignment.

---

*[src/core/core.cpp, line 1086 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIeG4S4UStHNe_imgH:-KZIeG4TQnkhqM2IF4eN:bt0qc4z) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1086)):*
> ```cpp
>     size_t length = tox_conference_peer_get_name_size(tox, groupId, peerId, &error);
>     if (!parsePeerQueryError(error))
>         return QString();
> ```

Prefer using `QString{};` or just `{};` to indicate an invocation of the `QString` no-arg/default constructor rather than an arbitary global function called `QString`. This goes for every instance replaced by this PR.

---

*[src/core/core.cpp, line 1133 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIevtyM_WTivJJ3njB:-KZIevtyM_WTivJJ3njC:b4270xb) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1133)):*
> ```cpp
>     }
> 
>     uint16_t nPeers = static_cast<uint16_t>(result);
> ```

Why is the number of peers represented as a unsigned 16-bit number here whereas the `getGroupNumberPeers()` returns a 32-bit number. Is it stated anywhere that peer numbers are 16-bits or less?

---

*[src/core/core.cpp, line 1134 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIi1mrWBdeKFvByvJb:-KZIi1mrWBdeKFvByvJc:b1tiaol) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1134)):*
> ```cpp
> 
>     uint16_t nPeers = static_cast<uint16_t>(result);
>     std::unique_ptr<uint8_t[][TOX_MAX_NAME_LENGTH]> namesArray{new uint8_t[nPeers][TOX_MAX_NAME_LENGTH]};
> ```

Could this not be refactored into using a std::vector? It seems to be much safer than manipulating multidimensional arrays within unique_ptrs.

---

*[src/core/core.cpp, line 1139 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIiIG-ZIhT_y2PsLjA:-KZIiIG-ZIhT_y2PsLjB:b4rn6bh) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1139)):*
> ```cpp
> 
>     uint32_t count = tox_conference_peer_count(tox, groupId, &error);
>     if (!parsePeerQueryError(error))
> ```

Missing braces around if-statement.

---

*[src/core/core.cpp, line 1245 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIiap9vGJsVVAieFlq:-KZIiapA-Q64ICoYvkPv:b-tgptxw) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1245)):*
> ```cpp
>         return;
>     case TOX_ERR_CONFERENCE_DELETE_CONFERENCE_NOT_FOUND:
>         qCritical() << "Conferenct not found";
> ```

Possible spelling mistake.

---

*[src/core/core.cpp, line 1265 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIifA92yYXVKiMBpKs:-KZIifA92yYXVKiMBpKt:bfqb7mx) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1265)):*
> ```cpp
>         break;
>     case TOX_ERR_CONFERENCE_INVITE_FAIL_SEND:
>         qCritical() << "Fail send";
> ```

I think this error message needs to be more descriptive.

---

*[src/core/core.cpp, line 1319 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIipLAPSpkFIqy0KcM:-KZIipLBLH-SzqpsGqQI:b-3oiivg) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1319)):*
> ```cpp
> {
>     // Valid length check
>     if (addr.length() != (TOX_ADDRESS_SIZE * 2))
> ```

Why was the braces removed around this if-statement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo committed:
refactor: rename args to conform to C++11 § 17.6.4.3.2

Fixed typo and small style improvements.

Diadlo commented:
Review status: 16 of 20 files reviewed at latest revision, 14 unresolved discussions.

---

*[src/core/core.cpp, line 507 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZISrAux5hg9wq6W1I4:-KZIkt4nfhZeGy5Lwaaz:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L507)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Pointer aligns to the left.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 507 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIVwtEC0_FgBlyIVjM:-KZIoJYb8085omg8NbXV:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L507)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

See point below on reserved identifiers.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 528 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIUULLG1X9vn90Douw:-KZIoKfcyKL2NgcE-3_Y:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L528)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

The use of an identifier who's name starts with an underscore is prohibited in C++:

As per § 17.6.4.3.2:

> Each name that begins with an underscore is reserved to the implementation for use as a name in the
global namespace.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 533 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIY_InHbw9GoJjB2Ww:-KZIoZBI8ox5m3U44Utf:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L533)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

```c++
bool isAction;
isAction = type == TOX_MESSAGE_TYPE_ACTION;
```

can be combined into 1 line.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 545 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIVSpn-sarMJQ5fANx:-KZIocaD9Tknn_F3rEEb:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L545)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

See point above.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 553 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIWAKM9_6hLJrU9sPq:-KZIosywEb1hJfZqBJ0W:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L553)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

You fixed pointer alignment above, but not here. Consider fixing this one too.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 646 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIWF5XxitLTTlNEvBj:-KZIp0DxSLIFCVmpIFBb:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L646)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Reference alignment.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1086 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIeG4S4UStHNe_imgH:-KZIpGpDOwQ_EZnBQTY6:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1086)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Prefer using `QString{};` or just `{};` to indicate an invocation of the `QString` no-arg/default constructor rather than an arbitary global function called `QString`. This goes for every instance replaced by this PR.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1133 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIevtyM_WTivJJ3njB:-KZIqZqpff0TMiPCKtI5:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1133)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Why is the number of peers represented as a unsigned 16-bit number here whereas the `getGroupNumberPeers()` returns a 32-bit number. Is it stated anywhere that peer numbers are 16-bits or less?
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1134 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIi1mrWBdeKFvByvJb:-KZIqYUCS_aQ8-2CcRs3:b-j1vdsl) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1134)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Could this not be refactored into using a std::vector? It seems to be much safer than manipulating multidimensional arrays within unique_ptrs.
</blockquote></details>

TODO added

---

*[src/core/core.cpp, line 1139 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIiIG-ZIhT_y2PsLjA:-KZIqaEZEqikb6sLvqMY:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1139)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Missing braces around if-statement.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1245 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIiap9vGJsVVAieFlq:-KZIr-8G3MRMXn9QKWl1:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1245)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Possible spelling mistake.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1265 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIifA92yYXVKiMBpKs:-KZIt-QH-KraJxDXr1Ut:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1265)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

I think this error message needs to be more descriptive.
</blockquote></details>

Done.

---

*[src/core/core.cpp, line 1319 at r10](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIipLAPSpkFIqy0KcM:-KZIrXaAsUVG9eLoIyv4:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/33050716547bdf88326cd7db712100f85028ca4f/src/core/core.cpp#L1319)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Why was the braces removed around this if-statement?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

initramfs commented:
Review status: 16 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/core/core.cpp, line 1152 at r11](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIwK8VSOWZVsRvCdYP:-KZIwK8W7HYBKZPyYrX6:bwy3lz0) ([raw file](https://github.com/qtox/qtox/blob/28f48534810fd9790ac639aedf51e0ddb7ff63d7/src/core/core.cpp#L1152)):*
> ```cpp
> 
>     QList<QString> names;
>     for (uint16_t i = 0; i < nPeers; ++i)
> ```

Since you've updated `nPeers` to a `std::uint32_t`, you need to update this for-loop temp variable type as well. A future-proof way is to declare is as a `decltype(nPeers)`, if not, simply change it to a `std::uint32_t`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Review status: 15 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/core/core.cpp, line 1152 at r11](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZIwK8VSOWZVsRvCdYP:-KZIyrtOLGScFidLziT_:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/28f48534810fd9790ac639aedf51e0ddb7ff63d7/src/core/core.cpp#L1152)):*
<details><summary><i>Previously, initramfs wrote…</i></summary><blockquote>

Since you've updated `nPeers` to a `std::uint32_t`, you need to update this for-loop temp variable type as well. A future-proof way is to declare is as a `decltype(nPeers)`, if not, simply change it to a `std::uint32_t`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

initramfs commented:
Reviewed 1 of 10 files at r4, 3 of 4 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/core/core.h, line 215 at r12](https://reviewable.io:443/reviews/qtox/qtox/3736#-KZJ0Jnf3fTstvklPq7i:-KZJ0Jnf3fTstvklPq7j:b7o0jmn) ([raw file](https://github.com/qtox/qtox/blob/4f9cb0b008eb91e3cadd164559486df16197d043/src/core/core.h#L215)):*
> ```objc
>     static void onConnectionStatusChanged(Tox* tox, uint32_t friendId,
>                                           TOX_CONNECTION status, void* core);
>     static void onGroupInvite(Tox* tox, uint32_t friendId, TOX_CONFERENCE_TYPE type,
> ```

The variable names changed in the function definition should be changed here in the declaration too. Or else you have inconsistent variable names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736)*
<!-- Sent from Reviewable.io -->

initramfs commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3736#-:-KZJ8m_diN6ZYLiESqo6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

zetok commented:
cc package maintainers: @abbat @farseerfc @mpxc @vith @yurivict

Heads up: with this qTox depends on c-toxcore, and 1.7 release (due in a week) will contain this change.

abbat commented:
Is there a way to compile qTox with original (iru) toxcore (through #define for example)? I have no motivation to support another (and conflicting) package which brings no new except breaking compatibility.

sudden6 commented:
@abbat toktok/c-toxcore doesn't break the protocol compatibility with irungentoo/toxcore

Additionally toktok/c-toxcore has active development, community and testing. All actively developed clients will make the switch, so I'd drop irungentoo/toxcore from the repo.
